proxy/http3/Http3Transaction.cc

```diff
@@ -234,24 +229,112 @@ HQTransaction::direction() const
   return this->_proxy_ssn->get_netvc()->get_context();
 }
 
-/**
- * @brief Replace existing event only if the new event is different than the inprogress event
- */
-Event *
-HQTransaction::_send_tracked_event(Event *event, int send_event, VIO *vio)
+void
+HQTransaction::_schedule_read_ready_event()
 {
-  if (event != nullptr) {
-    if (event->callback_event != send_event) {
-      event->cancel();
-      event = nullptr;
-    }
+  if (this->_read_ready_event) {
+    // The event is already scheduled. No need to schedule the same event.
+    return;
   }
 
-  if (event == nullptr) {
-    event = this_ethread()->schedule_imm(this, send_event, vio);
+  this->_read_ready_event = this->_thread->schedule_imm(this, VC_EVENT_READ_READY, &this->_read_vio);
+}
+
+void
+HQTransaction::_unschedule_read_ready_event()
+{
+  if (this->_read_ready_event) {
+    this->_read_ready_event->cancel();
+    this->_read_ready_event = nullptr;
   }
+}
 
-  return event;
+void
+HQTransaction::_close_read_ready_event(Event *e)
+{
+  ink_assert(this->_read_ready_event == e);
+  this->_read_ready_event = nullptr;
+}
+
+void
+HQTransaction::_schedule_read_complete_event()
+{
+  if (this->_read_complete_event) {
+    // The event is already scheduled. No need to schedule the same event.
+    return;
+  }
+
+  this->_read_complete_event = this->_thread->schedule_imm(this, VC_EVENT_READ_COMPLETE, &this->_read_vio);
+}
+
+void
+HQTransaction::_unschedule_read_complete_event()
+{
+  if (this->_read_complete_event) {
+    this->_read_complete_event->cancel();
+    this->_read_complete_event = nullptr;
+  }
+}
+
+void
+HQTransaction::_close_read_complete_event(Event *e)
+{
+  ink_assert(this->_read_complete_event == e);
+  this->_read_complete_event = nullptr;
+}
+
+void
+HQTransaction::_schedule_write_ready_event()
+{
+  if (this->_write_ready_event) {
+    // The event is already scheduled. No need to schedule the same event.
+    return;
+  }
+
+  this->_write_ready_event = this->_thread->schedule_imm(this, VC_EVENT_WRITE_READY, &this->_write_vio);
+}
+
+void
+HQTransaction::_unschedule_write_ready_event()
+{
+  if (this->_write_ready_event) {
+    this->_write_ready_event->cancel();
+    this->_write_ready_event = nullptr;
+  }
+}
+
+void
+HQTransaction::_close_write_ready_event(Event *e)
+{
+  ink_assert(this->_write_ready_event == e);
+  this->_write_ready_event = nullptr;
+}
+
+void
+HQTransaction::_schedule_write_complete_event()
+{
+  if (this->_write_complete_event) {
+    // The event is already scheduled. No need to schedule the same event.
+    return;
+  }
+
+  this->_write_complete_event = this->_thread->schedule_imm(this, VC_EVENT_WRITE_COMPLETE, &this->_write_vio);
+}
+
+void
+HQTransaction::_unschedule_write_complete_event()
+{
+  if (this->_read_complete_event) {
```

Comment:
shouldn't we be checking `_write_complete_event`?

Reply:
Yes. Thanks.